examples/applications/simple-provider/src/main.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+mod chariott_provider;
+
+use std::{net::SocketAddr, sync::Arc, time::Duration};
+
+use url::Url;
+
+use chariott_common::error::Error;
+use chariott_common::shutdown::RouterExt as _;
+use chariott_proto::{
+    provider::provider_service_server::ProviderServiceServer,
+    runtime::{
+        chariott_service_client::ChariottServiceClient, intent_registration::Intent,
+        intent_service_registration::ExecutionLocality, AnnounceRequest, IntentRegistration,
+        IntentServiceRegistration, RegisterRequest, RegistrationState,
+    },
+};
+use examples_common::chariott;
+use tokio::time::sleep;
+use tonic::transport::{Channel, Server};
+use tracing::warn;
+
+use crate::chariott_provider::ChariottProvider;
+
+#[derive(Clone)]
+struct RegisterParams {
+    name: String,
+    namespace: String,
+    version: String,
+    intents: Vec<Intent>,
+    url: String,
+    chariott_url: String,
+    locality: ExecutionLocality,
+}
+
+async fn connect_chariott_client(
+    client: &mut Option<ChariottServiceClient<Channel>>,
+    chariott_url: String,
+) -> Result<(), Error> {
+    *client = Some(ChariottServiceClient::connect(chariott_url).await.map_err(|e| {
+        *client = None; // Set client back to None on error.
+        Error::from_error("Could not connect to client", Box::new(e))
+    })?);
+
+    Ok(())
+}
+
+async fn register_announce_once(
+    client: &mut Option<ChariottServiceClient<Channel>>,
+    reg_params: RegisterParams,
+) -> Result<(), Error> {
+    // If there is no client, need to attempt connection.
+    if client.is_none() {
+        connect_chariott_client(client, reg_params.chariott_url).await?;
+    }
+
+    let service = Some(IntentServiceRegistration {
+        name: reg_params.name,
+        url: reg_params.url,
+        version: reg_params.version,
+        locality: reg_params.locality as i32,
+    });
+
+    let announce_req = AnnounceRequest { service: service.clone() };
+
+    // Allways announce to Chariott.
+    let registration_state = client
+        .as_mut()
+        .expect("No client found")
+        .announce(announce_req.clone())
+        .await
+        .map_err(|e| Error::from_error("Error announcing to Chariott.", Box::new(e)))?
+        .into_inner()
+        .registration_state;
+
+    // Only attempt registration with Chariott if the announced state is ANNOUNCED.
```

Comment:
Consider adding a comment to explain that getting "ANNOUNCED" state means that this service is not registered/ needs to register with Chariott. I know it is in the documentation elsewhere, but it could be helpful to have close to the code too.

Reply:
Updated to add info about the 'ANNOUNCED' state